Lecture 3 pdp add2cart

nika-doroshkevich opened the pull request.

nika-doroshkevich committed:
Task 3.1: CartService

nika-doroshkevich committed:
Task 3.2: PDP add2cart

nika-doroshkevich committed:
Task 3.3: HttpSessionCartService

nika-doroshkevich committed:
Task 3.4: Implement 3 recently viewed products

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,56 @@
+        for (CartItem cartItem : cart.getItems()) {
+            if (productId.equals(cartItem.getProduct().getId())) {
+                int previousQuantity = cartItem.getQuantity();
+                cartItem.setQuantity(previousQuantity + quantity);
```
What if new quantity is bigger than stock?

nika-doroshkevich replied:
Added

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,56 @@
+
+        boolean itemNotExists = true;
+
+        for (CartItem cartItem : cart.getItems()) {
```
Can you find existing item with selected product using streams?

nika-doroshkevich replied:
Yes

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:
```diff
@@ -173,6 +178,38 @@ public List<Product> findAllProducts() {
+        } else {
+            products = productsFromSession;
+            var firstProduct = products.getFirst();
+            if (!Objects.equals(firstProduct.getId(), product.getId())) {
```
What if requested product is in the list of recently viewed but not in first place?

nika-doroshkevich replied:
Fixed

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:
```diff
@@ -173,6 +178,38 @@ public List<Product> findAllProducts() {
+            if (!Objects.equals(firstProduct.getId(), product.getId())) {
+                products.addFirst(product);
+            }
+            if (products.size() > 3) {
```
Move recently view products list size to class constant

nika-doroshkevich replied:
Ok

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:
```diff
@@ -173,6 +178,38 @@ public List<Product> findAllProducts() {
+    }
+
+    @Override
+    public LinkedList<Product> getRecentlyViewedProducts(HttpServletRequest httpServletRequest) {
```
Move logic of recently viewed products to separate class

nika-doroshkevich replied:
Ok

AnnaKrikun commented on src/main/webapp/WEB-INF/pages/productList.jsp:
```diff
@@ -40,10 +40,37 @@
         </c:forEach>
     </table>
+
+    <p>
+        Recently viewed
+    </p>
+    <ul style="text-align: left; padding-inline-start: 0px;">
+        <c:forEach var="product" items="${recentlyViewedProducts}">
+            <li style="display: inline-block;">
+                <table>
+                    <tr>
+                        <td style="text-align: center;">
+                            <img class="product-tile" src="${product.imageUrl}">
+                            <div>
+                                <a href="${pageContext.servletContext.contextPath}/products/${product.id}">
+                                        ${product.description}
+                                </a>
+                            </div>
+                            <div>
+                                <fmt:formatNumber value="${product.price}" type="currency"
+                                                  currencySymbol="${product.currency.symbol}"/>
+                            </div>
+                        </td>
+                    </tr>
+                </table>
+            </li>
+        </c:forEach>
+    </ul>
```
Move to separate tag so it can be reused

nika-doroshkevich replied:
Ok

AnnaKrikun commented:
Add tests

nika-doroshkevich committed:
Added: recentlyViewedProducts tag, products list size constant, check when new quantity is bigger than stock
Fixed: find existing item with selected product using streams, bag about recently viewed products, move logic of recently viewed products to separate class

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/cart/DefaultCartService.java:
```diff
@@ -39,17 +39,19 @@ public synchronized void add(Cart cart, Long productId, int quantity) throws Out
+            var newQuantity = previousQuantity + quantity;
+            if (product.getStock() < newQuantity) {
+                throw new OutOfStockException(product, quantity, product.getStock());
             }
```
Move check to separate method to avoid code duplication

nika-doroshkevich replied:
Ok

AnnaKrikun commented on src/main/java/com/es/phoneshop/model/product/RecentlyViewedProductsService.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.LinkedList;
+
+public class RecentlyViewedProductsService {
```
Create RecentlyViewedProductsService interface and make class implement it

nika-doroshkevich replied:
Ok

AnnaKrikun commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -60,7 +61,7 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
             Cart cartFromSession = cartService.getCart(request);
             request.setAttribute("cart", cartFromSession);
-            productDao.addProductToRecentlyViewed(request, product);
+            new RecentlyViewedProductsService().addProductToRecentlyViewed(request, product);
```
Create field RecentlyViewedProductsService in Servlet class as productDao and CartService

nika-doroshkevich replied:
Ok

AnnaKrikun commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -39,7 +40,7 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
         request.setAttribute("products", products);
 
-        var list = productDao.getRecentlyViewedProducts(request);
+        var list = new RecentlyViewedProductsService().getRecentlyViewedProducts(request);
```
Create field RecentlyViewedProductsService in Servlet class

nika-doroshkevich replied:
Ok

AnnaKrikun commented on src/test/java/com/es/phoneshop/model/product/RecentlyViewedProductsServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RecentlyViewedProductsServiceTest {
```
Check formatting in class

nika-doroshkevich replied:
Checked
